Resolve type variable recursively in `GenericTypeResolver`

quaff opened the pull request:
- Addresses #24963

quaff committed:
Improve GenericTypeResolver to resolve type variable recursively

Fix GH-24963

quaff commented:
Rebased to main, ready for review.

snicoll commented on spring-core/src/main/java/org/springframework/core/GenericTypeResolver.java:
```diff
@@ -167,25 +168,32 @@ public static Type resolveType(Type genericType, @Nullable Class<?> contextClass
 					else if (typeArgument instanceof WildcardType wildcardType) {
-						generics[i] = resolveWildcard(wildcardType, contextType).resolve();
+						ResolvableType result = resolveWildcard(wildcardType, contextType);
+						if (result.resolve() == null) {
```
I wonder if that logic couldn't be applied in `resolveWildcard` since that's the only use of it. This would allow this to be a one liner like the other cases.

quaff replied:
> I wonder if that logic couldn't be applied in `resolveWildcard` since that's the only use of it. This would allow this to be a one liner like the other cases.

This change makes `GenericTypeResolverTests.resolveWildcardTypeWithUnbounded()` pass.

snicoll replied:
OK I am not saying removing. I am thinking about moving `result.resolve()` to the `resolveWildcard` method.

quaff replied:
OK, updated as you suggested.

snicoll reviewed (changes requested):
Thanks for the PR @quaff - This looks promising as it fixes another issue that was reported against generics handling.

Unfortunately, the build is broken with those changes, see https://ge.spring.io/s/kktsevdqhmgqc/tests/overview?outcome=FAILED.

quaff commented:
> Thanks for the PR @quaff - This looks promising as it fixes another issue that was reported against generics handling.
> 
> Unfortunately, the build is broken with those changes, see https://ge.spring.io/s/kktsevdqhmgqc/tests/overview?outcome=FAILED.

The failed tests should adapt changes on [ResolvableType.java](https://github.com/spring-projects/spring-framework/pull/30079/files#diff-58e4c9d4400aa89a869df68b7c8974fe646147a93e41dc7c1535306996e5f550), I've updated the commit.

snicoll commented:
> The failed tests should adapt changes on [ResolvableType.java](https://github.com/spring-projects/spring-framework/pull/30079/files#diff-58e4c9d4400aa89a869df68b7c8974fe646147a93e41dc7c1535306996e5f550), I've updated the commit.

With those changes in place, you had to change some assertions that looks invalid to me now. For instance, event listening that accepts a `String` as input would be considered if the type is `EntityWrapper<?>`. Same for an event that accepts an `EntityWrapper<?>` would now accept a `List<?>`. 

This may be something that wrong elsewhere that the fix reveals but changing the assertion isn't what I had in mind.

Given the impact this has so late, I've moved this one for 6.2. If you have time to review the above, that would be much appreciated.

quaff commented:
> > The failed tests should adapt changes on [ResolvableType.java](https://github.com/spring-projects/spring-framework/pull/30079/files#diff-58e4c9d4400aa89a869df68b7c8974fe646147a93e41dc7c1535306996e5f550), I've updated the commit.
> 
> With those changes in place, you had to change some assertions that looks invalid to me now. For instance, event listening that accepts a `String` as input would be considered if the type is `EntityWrapper<?>`. Same for an event that accepts an `EntityWrapper<?>` would now accept a `List<?>`.
> 
> This may be something that wrong elsewhere that the fix reveals but changing the assertion isn't what I had in mind.
> 
> Given the impact this has so late, I've moved this one for 6.2. If you have time to review the above, that would be much appreciated.

I've fixed that, please review. @snicoll

quaff commented:
Any chance to bring this to 6.1? @jhoeller @snicoll

snicoll commented:
@quaff please refrain from ping like this. You've already ping me once and trust that we're doing our best to handle it in time if we can.

snicoll commented:
@quaff with https://github.com/spring-projects/spring-framework/issues/31690#issuecomment-1831867944 some code that we didn't intend to have has been reverted. If you have time to rebase, it would be much appreciated.

quaff commented:
> @quaff with [#31690 (comment)](https://github.com/spring-projects/spring-framework/issues/31690#issuecomment-1831867944) some code that we didn't intend to have has been reverted. If you have time to rebase, it would be much appreciated.

Done.

kilink commented:
We happened to be testing our apps on the latest Spring snapshot, and I believe this change has introduced an infinite recursion bug which causes a StackOverFlowError. When we start several of our Spring Boot apps, we end up with the following error:
```
Exception in thread "main" java.lang.StackOverflowError
	at org.springframework.util.ObjectUtils.nullSafeHashCode(ObjectUtils.java:451)
	at org.springframework.core.ResolvableType.calculateHashCode(ResolvableType.java:1012)
	at org.springframework.core.ResolvableType.hashCode(ResolvableType.java:1000)
	at org.springframework.util.ObjectUtils.nullSafeHashCode(ObjectUtils.java:451)
	at org.springframework.core.ResolvableType.calculateHashCode(ResolvableType.java:1012)
	at org.springframework.core.ResolvableType.hashCode(ResolvableType.java:1000)
	at org.springframework.util.ObjectUtils.nullSafeHashCode(ObjectUtils.java:451)
```

I narrowed it down to a specific ApplicationEvent which triggers it. The following will reproduce the problem:

```java
AvailabilityChangeEvent<LivenessState> event = new AvailabilityChangeEvent<>(new Object(), LivenessState.CORRECT);
ResolvableType resolvableType = ResolvableType.forInstance(event);
resolvableType.hasUnresolvableGenerics();
```
[AvailabilityChangeEvent](https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/availability/AvailabilityChangeEvent.java) and [LivenessState](https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/availability/LivenessState.java) are both classes from Spring Boot.

kilink commented:
Looks like it can be reproduced directly with just an Enum:

```java
enum Foo { BAR }
ResolvableType.forType(Foo.class.getGenericSuperclass()).hasUnresolvableGenerics();
// or
ResolvableType.forClassWithGenerics(Enum.class, Foo.class).hasUnresolvableGenerics();
```

kilink commented:
Seems like it's self-bounding generics that are the problem, the following triggers the bug as well:

```java
class Foo<T extends Foo<T>> {}
class Bar extends Foo<Bar> {}

ResolvableType.forInstance(new Bar()).hasUnresolvableGenerics();
```
